spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/ChangelogIterator.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * <p>It marks the carry-over rows to null to for filtering out later. Carry-over rows are unchanged
+ * rows in a snapshot but showed as delete-rows and insert-rows in a changelog table due to the
+ * copy-on-write(COW) mechanism. For example, there are row1 (id=1, data='a') and row2 (id=2,
+ * data='b') in a data file, if we only delete row2, the COW will copy row1 to a new data file and
+ * delete the whole old data file. The changelog table will have two delete-rows(row1 and row2), and
+ * one insert-row(row1). Row1 is a carry-over row.
+ *
+ * <p>The iterator marks the delete-row and insert-row to be the update-rows. For example, these two
+ * rows
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class ChangelogIterator implements Iterator<Row>, Serializable {
+  private static final String DELETE = ChangelogOperation.DELETE.name();
+  private static final String INSERT = ChangelogOperation.INSERT.name();
+  private static final String UPDATE_BEFORE = ChangelogOperation.UPDATE_BEFORE.name();
+  private static final String UPDATE_AFTER = ChangelogOperation.UPDATE_AFTER.name();
+
+  private final Iterator<Row> rowIterator;
+  private final int changeTypeIndex;
+  private final List<Integer> partitionIdx;
+
+  private Row cachedRow = null;
+
+  public ChangelogIterator(
+      Iterator<Row> rowIterator, int changeTypeIndex, List<Integer> partitionIdx) {
+    this.rowIterator = rowIterator;
+    this.changeTypeIndex = changeTypeIndex;
+    this.partitionIdx = partitionIdx;
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (cachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    // if there is an updated cached row, return it directly
+    if (updated(cachedRow)) {
+      Row row = cachedRow;
+      cachedRow = null;
+      return row;
+    }
+
+    Row currentRow = currentRow();
+
+    if (rowIterator.hasNext()) {
+      GenericRowWithSchema nextRow = (GenericRowWithSchema) rowIterator.next();
+      cachedRow = nextRow;
+
+      if (updateOrCarryoverRecord(currentRow, nextRow)) {
+        Row[] rows = update((GenericRowWithSchema) currentRow, nextRow);
+
+        currentRow = rows[0];
+        cachedRow = rows[1];
+      }
+    }
+
+    return currentRow;
+  }
+
+  private Row[] update(GenericRowWithSchema currentRow, GenericRowWithSchema nextRow) {
+    GenericInternalRow deletedRow = new GenericInternalRow(currentRow.values());
+    GenericInternalRow insertedRow = new GenericInternalRow(nextRow.values());
+
+    if (isCarryoverRecord(deletedRow, insertedRow)) {
+      // set carry-over rows to null for filtering out later
+      return new Row[] {null, null};
```

Comment:
Curious question, is the null filter in Spark later, just for this case?  Would there be some perf advantage do you think of taking care of this in the iterator, something like (very pseudo):

do {
   rows = nextInternal();
} while (row[0] != null and row[1] != null)

Reply:
We will filter out the null row by concatenating this iterator with another one like what we did in the test.
```
Iterators.filter(iterator, Objects::nonNull)
``` 

Here is another code example, https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L476-L476

I'm fine with either one, but the iterator concatenation would be a nice solution to reduce the complexity of this iterator.

Reply:
Right.  Maybe we can have this iterator just wrap Iterators.filter(...)?  If this is kind of a standalone class, not sure if there is anyone is interested in having nulls?  wdyt?

Reply:
I was planning to concatenate them in the CDC procedure. I don't think anyone would be interested in the nulls. Let me think about how to merge them within a class.

Reply:
This seems like a very expensive way to pass back 2 nulls

Reply:
I would probably just return "null" and do an "if null { currentRow = null , cachedRow = null }

Reply:
Was trying to make it more readable by return 2 nulls. Will make the change.

Reply:
A static method is used for concatenating two iterators.